### PR TITLE
Max7219digit multiline

### PR DESCRIPTION
**Pull request by TVDLoewe**
# What does this implement/fix? 

Adds multi line functionality. All new parameters are optional and with default values the behaviour should not change.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (this will require users to update their yaml configuration files to keep working)

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>
  
# Test Environment

- [x] ESP32
- [ ] ESP8266
- [x] Windows
- [ ] Mac OS
- [ ] Linux

## Example entry for `config.yaml`:

```yaml
esphome:
  name: led_display
  platform: ESP32
  board: nodemcu-32s

wifi:
  ssid: "<your ssid>"
  password: "<your pw>"

# Enable Home Assistant API
api: 
  password: "<your pw>"

ota: 
  password: "<your pw>"

logger:

spi:
  clk_pin: GPIO013 
  mosi_pin: GPIO014

display:
  - platform: max7219digit
    cs_pin: GPIO015
    num_chips: 20
    num_chip_lines: 2 # default is 1                  --> new parameter
#    chip_lines_style: zigzag # default is snake      --> new parameter
    scroll_mode: stop
#    reverse_enable: true
#    rotate_chip: 180
#    scroll_speed: 2s
    intensity: 1
    lambda: |-
      it.printf(0, 0, id(digit_font), id(hatxt).state.c_str());
      it.intensity(atoi(id(habri).state.c_str()));

font:
  - file: "pixelmix.ttf"
    id: digit_font
    size: 16 #          --> i am using 2 chip_lines so my display is 16 LEDs high
    
text_sensor:
  - platform: homeassistant
    name: "Display Text"
    id: hatxt
    entity_id: input_text.display_text
  - platform: homeassistant
    name: "Display Brightness"
    id: habri
    entity_id: input_number.display_bri

```

# Explain your changes
IMG_20210319_180150

With this new feature you can stack modules like "MAX7219 8x32 4 in 1 Dot Matrix" on top of another and get a bigger display this way. I am using this feature as a custom component and may be some one else would like this as well.

I am not very familiar with github but maybe you can help me figuring the rest out.

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/max7219digit/display.py [hunk 1]
```diff
@@ -13,6 +13,13 @@
 CONF_SCROLL_ENABLE = "scroll_enable"
 CONF_SCROLL_MODE = "scroll_mode"
 CONF_REVERSE_ENABLE = "reverse_enable"
+CONF_NUM_CHIP_LINES = "num_chip_lines"
+CONF_CHIP_LINES_STYLE = "chip_lines_style"
+
+CHIP_LINES_STYLE = {
+    "ZIGZAG": 0,
```
Comment (OttoWinter) [inline esphome/components/max7219digit/display.py:20]: This should also be an enum in the C++ code (declare an `enum` block there with constants for STYLE_ZIGZAG and STYLE_SNAKE).

See https://github.com/esphome/esphome/blob/91898cb814332936d452599070e62cb380e519ee/esphome/components/integration/integration_sensor.h#L11-L17 for an example

File: esphome/components/max7219digit/max7219digit.cpp [hunk 2]
```diff
@@ -25,9 +25,12 @@ void MAX7219Component::setup() {
...
   this->spi_setup();
   this->stepsleft_ = 0;
-  this->max_displaybuffer_.reserve(500);  // Create base space to write buffer
-  // Initialize buffer with 0 for display so all non written pixels are blank
-  this->max_displaybuffer_.resize(this->num_chips_ * 8, 0);
+  for (int chip_line = 0; chip_line < this->num_chip_lines_; chip_line++) {
+    std::vector<uint8_t> vec(1);
+    this->max_displaybuffer_.push_back(vec);
+    // Initialize buffer with 0 for display so all non written pixels are blank
+    this->max_displaybuffer_[chip_line].resize(get_width_internal(), 0);
```
Comment (OttoWinter) [inline esphome/components/max7219digit/max7219digit.cpp:32]: This block can be simplified to:
```cpp
this->max_displaybuffer_ = {num_chip_lines_, {get_width_internal(), 0}};
```

**Review by OttoWinter — APPROVED** [review]
Thanks! Looks good to me in general, left some comments on minor code improvements for you.

**Comment by rspaargaren** [comment]
Does this work with the scroll mode?

**Comment by rspaargaren** [comment]
And the rotated configuration?

File: esphome/components/max7219digit/max7219digit.cpp (hunk 2 above)
Comment (TVDLoewe) [reply esphome/components/max7219digit/max7219digit.cpp:32]: I could not get this working.

**Comment by TVDLoewe** [comment]
Yes, scroll mode is working.

If you set reverse_enable to true and rotate_chip to 180 and switch the chip lines the display is on its head.

**Comment by TVDLoewe** [comment]
Hello @rspaargaren and @OttoWinter, some time has passed without any change in this pull request. I am not sure what the normal procedure is here. Is there anything, i can do to help you?

**Comment by OttoWinter** [comment]
@TVDLoewe Looks ok to me from the code side. But I can't comment on if this might cause any issues with how max7219digit works internally. Ok to merge if @rspaargaren agrees :+1:

**Comment by TVDLoewe** [comment]
Hey @rspaargaren any news on this issue from your side?

**Comment by jesserockz** [comment]
@TVDLoewe If you can add this option to the docs today I can merge this for the next release 18 hours from now (provided CI passes still)

**Comment by rspaargaren** [comment]
> @TVDLoewe If you can add this option to the docs today I can merge this for the next release 18 hours from now (provided CI passes still)

I have missed out on this one will have a quick look as well

**Comment by TVDLoewe** [comment]
Hey @jesserockz, i have added some Doku in https://github.com/esphome/esphome-docs/pull/1602

**Pull request merged**